src/internal/internal_gemm.cc

```diff
@@ -479,136 +479,18 @@ void gemm(internal::TargetType<Target::Devices>,
-                        }
-                    }
-                }
-            }
-
-            // right column
-            std::vector<scalar_t*> a_array01;
-            std::vector<scalar_t*> b_array01;
-            std::vector<scalar_t*> c_array01;
-            a_array01.reserve( batch_size );
-            b_array01.reserve( batch_size );
-            c_array01.reserve( batch_size );
-
-            int64_t lda01 = 0;
-            int64_t ldb01 = 0;
-            int64_t ldc01 = 0;
-            int64_t mb01 = C.tileMb(0);
-            int64_t nb01 = C.tileNb(C.nt()-1);
-            // same kb as above
-            {
-                int64_t j = C.nt()-1;
-                for (int64_t i = 0; i < C.mt()-1; ++i) {
-                    if (C.tileIsLocal(i, j)) {
-                        if (device == C.tileDevice(i, j)) {
-                            a_array01.push_back( A(i, 0, device).data() );
-                            b_array01.push_back( B(0, j, device).data() );
-                            c_array01.push_back( C(i, j, device).data() );
-                            lda01 = A(i, 0, device).stride();
-                            ldb01 = B(0, j, device).stride();
-                            ldc01 = C(i, j, device).stride();
-                        }
-                    }
-                }
-            }
+            scalar_t** a_array_host = C.array_host(device, queue_index);
+            scalar_t** b_array_host = a_array_host + batch_size;
+            scalar_t** c_array_host = b_array_host + batch_size;
 
-            // bottom-right corner
-            std::vector<scalar_t*> a_array11;
-            std::vector<scalar_t*> b_array11;
-            std::vector<scalar_t*> c_array11;
-
-            int64_t lda11 = 0;
-            int64_t ldb11 = 0;
-            int64_t ldc11 = 0;
-            int64_t mb11 = C.tileMb(C.mt()-1);
-            int64_t nb11 = C.tileNb(C.nt()-1);
-            // same kb as above
-            {
-                int i = C.mt()-1;
-                int j = C.nt()-1;
-                if (C.tileIsLocal(i, j)) {
-                    if (device == C.tileDevice(i, j)) {
-                        a_array11.push_back( A(i, 0, device).data() );
-                        b_array11.push_back( B(0, j, device).data() );
-                        c_array11.push_back( C(i, j, device).data() );
-                        lda11 = A(i, 0, device).stride();
-                        ldb11 = B(0, j, device).stride();
-                        ldc11 = C(i, j, device).stride();
-                    }
-                }
-            }
+            // C comes first since we do computation for a local C
+            auto group_params = device_regions_build<false, 3, scalar_t>(
+                                                    {C, A, B},
+                                                    {c_array_host, a_array_host, b_array_host},
+                                                    device );
```

Comment:
To avoid running over 80 chars, I would just indent this 4, or possibly 8, extra spaces, e.g.,
```
            // C comes first since we do computation for a local C
            auto group_params = device_regions_build<false, 3, scalar_t>(
                {C, A, B},
                {c_array_host, a_array_host, b_array_host},
                device );
```